TileStache/Goodies/VecTiles/transform.py

```diff
@@ -366,3 +366,90 @@ def tags_name_i18n(shape, properties, fid, zoom):
             properties[alt_tag_name_candidate] = alt_tag_name_value
 
     return shape, properties, fid
+
+# intercut takes features from a base layer and cuts each
+# of them against a cutting layer, splitting any base
+# feature which intersects into separate inside and outside
+# parts.
+#
+# the parts of each base feature which are outside any
+# cutting feature are left unchanged. the parts which are
+# inside have their property with the key given by the
+# 'target_attribute' parameter set to the same value as the
+# property from the cutting feature with the key given by
+# the 'attribute' parameter.
+#
+# the intended use of this is to project attributes from one
+# layer to another so that they can be styled appropriately.
+#
+# returns a set of feature layers with the base layer
+# replaced by a cut one, or None if there's an error.
+def intercut(feature_layers, base_layer, cutting_layer, attribute=None, target_attribute=None):
+    base = None
+    cutting = None
+
+    # the target attribute can default to the attribute if
+    # they are distinct. but often they aren't, and that's
+    # why target_attribute is a separate parameter.
+    if target_attribute is None:
+        target_attribute = attribute
+
+    # search through all the layers and extract the ones
+    # which have the names of the base and cutting layer.
+    # it would seem to be better to use a dict() for
+    # layers, and this will give odd results if names are
+    # allowed to be duplicated.
+    for feature_layer in feature_layers:
+        layer_datum = feature_layer['layer_datum']
+        layer_name = layer_datum['name']
+
+        if layer_name == base_layer:
+            base = feature_layer
+        elif layer_name == cutting_layer:
+            cutting = feature_layer
+
+    # didn't find one or other layer - what's the appropriate
+    # thing to do here, raise an error?
+    if base is None or cutting is None:
+        return None
+
+    base_features = base['features']
+    cutting_features = cutting['features']
+
+    # TODO: this is a very simple way of doing this, and would
+    # probably be better replaced by something that isn't O(N^2)
+    # and perhaps even unioned features with the same attribute
+    # together first.
```

Comment:
What I was thinking was that we want to minimise the number of cuts, as each split road is going to have overhead associated with having its own attributes, arcs, etc... Given that all we are interested in, for the purposes of splitting, is the attribute on the cutting layer then perhaps it makes sense to union together all the features in the cutting layer which share the same attribute value. Clipping against each of those, from highest priority to lowest, and not re-clipping the parts which are inside higher priority polygons should give us the minimum number of cuts.

Starting with (please excuse the ASCII art...):

```
  aaaaaaa
  a     a
  a   bbabbbb
  a   b a   b
  a ccbcacc b
--a-c-b-a-c-b--
  aaaaaaa c b
    c b   c b
    c bbbbbbb
    c     c
    ccccccc
```

Where the features `[a, b, c]` are in the cutting layer, and `a` and `c` have the same attribute, which is lower priority than `b`'s attribute. A line going through the center `---` would be cut into 7 pieces, and we've done 7 intersection operations. If we union together `a` and `c` and cut first against `b`, then we get:

```
  aaaaaaa
  a     a
  a   bbabbbb
  a   b a   b
  a   b aaa b
--a---b=====b--
  aaa b   a b
    a b   a b
    a bbbbbbb
    a     a
    aaaaaaa
```

By first cutting against the highest priority polygon, `b`, and setting aside the inside section (marked `===`) as it can't be cut by anything higher priority, and then the union of `a` and `c`, we get 4 sections instead of 7, and we've done 2 intersection operations.

Creating spatial indexes is an additional potential optimisation, as is using shapely's "prepared geometries", and I wouldn't want to guess what the impact would be of using either of these without trying to measure it first...